iou rest service

isfomin opened the pull request:
Решение задания 5. https://t.me/try2py/197

Добавлены:
- rest-controllers;
- services;
- entities;
- dto;
- converters;
- configuration;
- liquibase changelog;
- unit tests.

[Примеры и описание API](https://github.com/iSFomin/IouRestService/blob/master/README.md)

isfomin committed:
iou rest service is ready
added:
- rest-controllers;
- services;
- entities;
- dto;
- converters;
- configuration;
- liquibase changelog;
- unit tests.

iSFomin commented:
Что отправлять в ответ на операции добавления (POST)...
Id добавленной записи?
Добавленную сущность целиком (в моей реализации)?
Код 200 с текстом в теле что-то типа "добавлено успешно"?
Код 200 с пустым ответом?
Есть общие стандарты или правила хорошего тона или всё зависит от контекста задачи?

isfomin commented on src/main/kotlin/com/codereview/iou/service/PromissoryNoteService.kt:
```diff
@@ -0,0 +1,95 @@
+
+        val grouping = list
+            .groupingBy { "${it.lender?.id}-${it.borrower?.id}" }
+            .reduce { _, accumulator, element ->
```
плохо представляю как работает aggregate функция
хотелось бы reduce и map объединить во что-то... с помощью aggregate возможно так сделать?

iSFomin replied:
хотелось бы на примерах увидеть как можно сократить количество операций в этом коде

YourCodeReview commented:
Ссылка на задание:

https://t.me/try2py/197

isfomin commented on src/main/kotlin/com/codereview/iou/service/PromissoryNoteService.kt:
```diff
@@ -0,0 +1,95 @@
+            amount = promissoryNoteDto.amount
+        }
+
+        if (validatePromissoryNoteEntity(promissoryNote)) {
```
данные полученные от пользователя правильнее валидировать в объекте dto или в entity?

MardokUA replied:
К сежелению, я не back-end'щик и поэтому сказать как правильно с точки зрения такого разработчика я не могу. Однако, с точки зрения Clean Architecture, у меня ответ такой:
- если `PromissoryNoteService` часть `domain`слоя, то перед мапингом
- если часть `data` слоя, то вообще не здесь, а в `domain` слое

isfomin replied:
На самом деле я тоже не бэкенд разработчик, практикую котлин)
Clean Architecture да, спасибо, попробую углубится в этот вопрос.

MardokUA commented on build.gradle.kts:
```diff
@@ -16,10 +17,18 @@ repositories {
 
 dependencies {
-	implementation("org.springframework.boot:spring-boot-starter")
+	implementation("org.springframework.boot:spring-boot-starter:2.5.6")
```
я бы советовал выносить общие версии зависиомтсей в переменные. 
```suggestion
        val spring = "2.5.6"
	implementation("org.springframework.boot:spring-boot-starter:$spring")
```

MardokUA commented on src/main/kotlin/com/codereview/iou/controller/IouController.kt:
```diff
@@ -0,0 +1,43 @@
+class IouController {
+
+    @Autowired
+    lateinit var service: PromissoryNoteService
```
Я не знаком с Spring и как он доставляет зависимости, но, если можно, я бы организовывал DI в конструктор а не в свойство класса. Самый большой плюс - в тестах можно отказаться от фреймворка и самому мокать классы

MardokUA commented on src/main/kotlin/com/codereview/iou/controller/IouController.kt:
```diff
@@ -0,0 +1,43 @@
+    lateinit var service: PromissoryNoteService
+
+    @GetMapping
+    fun getAllIou(): PromissoryNotesDto {
```
Более идиомично писать вот так:
```suggestion
    fun getAllIou(): PromissoryNotesDto = service.listPromissoryNotes()
```

MardokUA commented on src/main/kotlin/com/codereview/iou/controller/IouController.kt:
```diff
@@ -0,0 +1,43 @@
+
+    @DeleteMapping
+    fun deleteUser(@RequestBody promissoryNoteDto: PromissoryNoteDto): ResponseEntity<MessageDto> {
+        service.deletePromissoryNote(promissoryNoteDto)
```
Вот тут не совсем понятно - а чтоесли удалить не получиться? Все равно нуно возвращать success?

isfomin replied:
На самом деле меня бы тоже смутил этот код. В спринге есть Exception Handler, если вдруг что-то пойдёт не так при удалении - выкинется исключение, отработает Exception Handler и ветка исполнения пойдет мимо "success".

iSFomin replied:
Так я думал что работает...
Но пытаюсь проверить и ведь реально "success" во всех случаях, кроме id = null который я явно обрабатываю

MardokUA commented on src/main/kotlin/com/codereview/iou/controller/UserController.kt:
```diff
@@ -0,0 +1,40 @@
+
+    @DeleteMapping
+    fun deleteUser(@RequestBody userDto: UserDto): ResponseEntity<MessageDto> {
+        service.deleteUser(userDto)
```
Тот же самый вопрос. Я поддерживаю следующих подход - если функция выполняет действие над базой данных, то всегда должен быть тип возвращаемого значение. Например, для вставки это `Int` в качестве количесва записей или массив  `long`в качестве массива IDшников. Тут вероятно подойдет даже `boolean`

isfomin replied:
spring data jpa для методов delete возвращает void, поэтому приходится полагаться на эксепшены
заменить текст "success" на количество удаленных записей, возможно да, так будет более информативно

MardokUA commented on src/main/kotlin/com/codereview/iou/controller/handler/RestResponseEntityExceptionHandler.kt:
```diff
@@ -0,0 +1,29 @@
+    @ExceptionHandler(IllegalArgumentException::class)
+    fun handleIllegalArgumentException (ex: Exception, request: WebRequest): ResponseEntity<ExceptionDto> {
+        return ResponseEntity<ExceptionDto>(
+            ExceptionDto(ex.message ?: ex.toString()), HttpHeaders(), HttpStatus.BAD_REQUEST
```
Правильный подход работы с ошибками, когда нужно обернуть ошибку системы в свою, это положить реальную ошибку в `cause` своей кастомной ошибки. Я понимаю, что Вы возвращаете DTO, однако держа все вышесказанное в голове, я бы в текст ошибки ложил бы `ex.message() ?: ex?.cause?.message ?: ex.toString()`

MardokUA commented on src/main/kotlin/com/codereview/iou/model/dto/ExceptionDto.kt:
```diff
@@ -0,0 +1,3 @@
+package com.codereview.iou.model.dto
+
+class ExceptionDto(val error: String)
```
Советовал бы для такой задачи использовать `inline/value class`, в зависимости от версии Kotlin

MardokUA commented on src/main/kotlin/com/codereview/iou/model/dto/PromissoryNoteCreateDto.kt:
```diff
@@ -0,0 +1,8 @@
+package com.codereview.iou.model.dto
+
+class PromissoryNoteCreateDto() {
```
Есть ли смысл использовать `data class`?

MardokUA commented on src/main/kotlin/com/codereview/iou/model/dto/UserDto.kt:
```diff
@@ -0,0 +1,6 @@
+package com.codereview.iou.model.dto
+
+class UserDto() {
```
Мне кажется, что все `DTO` классы можно смело делать `data class`

MardokUA commented on src/main/kotlin/com/codereview/iou/model/entity/PromissoryNote.kt:
```diff
@@ -0,0 +1,22 @@
+class PromissoryNote {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(nullable = false)
```
Повторюсь, что не знаком со Spring, но выглядит очень комично - nullable поле которое не может быть null)

MardokUA commented on src/main/kotlin/com/codereview/iou/service/PromissoryNoteService.kt:
```diff
@@ -0,0 +1,95 @@
+class PromissoryNoteService {
+
+    @Autowired
+    lateinit var iouRepository: PromissoryNoteRepository
```
опять же вопрос о `lateinit` - я предпочитаю injection в конструктор а не в поля. `lateinit` вообще считается огромной ошибкой в Kotlin. Это говорил сам Андрей Бреслав и для этих целей есть` Delegates.notNull()`

MardokUA commented on src/main/kotlin/com/codereview/iou/service/PromissoryNoteService.kt:
```diff
@@ -0,0 +1,95 @@
+    val converter: PromissoryNoteConverter = Mappers.getMapper(PromissoryNoteConverter::class.java)
+
+    fun addPromissoryNote(promissoryNoteDto: PromissoryNoteCreateDto): PromissoryNote {
+        val promissoryNote = PromissoryNote().apply {
```
почему бы сразу не создать инстас класса и не положить все в конструктор? Теперь я понимаю зачем Вам nullable поля. Я все же настаиваю на `data class` и `not null property` внутри. В котлине есть именованные аргументы, что позволяет избежать проблемы изменения положения свойств в конструкторе. Пологаю, именно с этой целью вы используете `apply` как замену паттерна `Builder`

isfomin replied:
1. data class не рекомендуется устанавливать для JPA сущности. Предупреждение от идеи: The data class implementations of equals(), hashCode() and toString() are not recommended for JPA entities. They can cause severe performance and memory consumption issues. 
2. перенести параметры в конструктор, в общем то ничего не мешает, можно попробовать

MardokUA commented on src/main/kotlin/com/codereview/iou/service/UserService.kt:
```diff
@@ -0,0 +1,53 @@
+        if (validateUserEntity(user)) {
+            return repository.save(user)
+        } else {
+            throw IllegalArgumentException("Entity {$user} not valid.")
```
Я бы советовал написать inline функцию, что-то вроде `validate(block: () -> Boolean, exp: <тип ошибки>})` и перегрузил бы его.

isfomin replied:
Хорошее замечание, учту)

MardokUA commented on src/main/kotlin/com/codereview/iou/service/UserService.kt:
```diff
@@ -0,0 +1,53 @@
+
+    fun getUser(userId: Long): UserDto {
+        val user = repository.findById(userId).get()
+        return converter.convertToDto(user)
```
```suggestion
        return epository.findById(userId).get().run {converter.convertToDto(this)}
```

MardokUA commented on src/test/kotlin/com/codereview/iou/service/PromissoryNoteServiceTest.kt:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Test
+    fun getNotExistPromissoryNote() {
```
Советую испоьзовать кричащие наименования функций и использовать Геркин (https://ru.wikipedia.org/wiki/BDD_(%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5))
пример, как переименовать этот тест: 
```suggestion
    fun `when note id is invalid, error should be returned`() {
```

isfomin replied:
Впервые об этом слышу, спасибо)

isfomin committed:
several improvements
added:
- unit test for controllers;
- test utils;
- Gherkin description test;
- upgrade validation;
- extensions;
- fixed errors.

isfomin committed:
several improvements #2
changes:
- di moved to primary constructor;
- use variable with version for group dependencies;
- other.

isfomin committed:
Apply suggestions from code review

Co-authored-by: Roman Laktionov <r.a.laktionov@gmail.com>

isfomin committed:
several improvements #3
changes:
- syntax fixed after review code;
- other

isfomin committed:
several improvements #4
changes:
- use non-null types in entity;
- other.
